pkg/controller/clone_controller_test.go

```diff
@@ -0,0 +1,226 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/informers"
+	fakeclientset "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/util/workqueue"
+)
+
+var requestedBytes int64 = 1000
+var fakeSc1 string = "fake-sc-1"
+
+const (
+	srcName      = "clone-source-pvc"
+	dstName      = "destination-pvc"
+	srcNamespace = "fake-pvc-namespace"
+	pvName       = "test-testi"
+)
+
+func pvcFinalizers(pvc *v1.PersistentVolumeClaim, finalizers ...string) *v1.PersistentVolumeClaim {
+	pvc.Finalizers = append(pvc.Finalizers, finalizers...)
+	return pvc
+}
+
+func pvcDataSourceClone(pvc *v1.PersistentVolumeClaim, srcName string) *v1.PersistentVolumeClaim {
+	apiGr := ""
+	pvc.Spec.DataSource = &v1.TypedLocalObjectReference{
+		APIGroup: &apiGr,
+		Kind:     pvcKind,
+		Name:     srcName,
+	}
+	return pvc
+}
+
+func pvcNamed(pvc *v1.PersistentVolumeClaim, name string) *v1.PersistentVolumeClaim {
+	pvc.Name = name
+	return pvc
+}
+
+func pvcNamespaced(pvc *v1.PersistentVolumeClaim, namespace string) *v1.PersistentVolumeClaim {
+	pvc.Namespace = namespace
+	return pvc
+}
+
+func pvcPhase(pvc *v1.PersistentVolumeClaim, phase v1.PersistentVolumeClaimPhase) *v1.PersistentVolumeClaim {
+	pvc.Status.Phase = phase
+	return pvc
+}
+
+func baseClaim() *v1.PersistentVolumeClaim {
+	return fakeClaim(srcName, srcNamespace, "fake-claim-uid", requestedBytes, pvName, v1.ClaimBound, &fakeSc1, "").DeepCopy()
+}
+
+func pvcDeletionMarked(pvc *v1.PersistentVolumeClaim) *v1.PersistentVolumeClaim {
+	timeX := metav1.NewTime(time.Now())
+	pvc.DeletionTimestamp = &timeX
+	return pvc
+}
+
+// TestCloneFinalizerRemoval tests create volume clone
+func TestCloneFinalizerRemoval(t *testing.T) {
+	testcases := map[string]struct {
+		initialClaims   []runtime.Object
+		cloneSource     runtime.Object
+		expectFinalizer bool
+		expectError     error
+	}{
+		"delete source pvc with no cloning in progress": {
+			cloneSource: pvcFinalizers(baseClaim(), pvcCloneFinalizer),
+			initialClaims: []runtime.Object{
+				pvcDataSourceClone(
+					pvcNamed(baseClaim(), dstName),
+					srcName,
+				),
+			},
+		},
+		"delete source pvc when destination pvc status is claim pending": {
+			cloneSource: pvcFinalizers(baseClaim(), pvcCloneFinalizer),
+			initialClaims: []runtime.Object{
+				pvcPhase(
+					pvcDataSourceClone(
+						pvcNamed(baseClaim(), dstName),
+						srcName,
+					),
+					v1.ClaimPending)},
```

Comment:
It would be more readable if you put `claim` as the last argument - function name + its arguments would be on the same line:

```
pvcPhase(v1.ClaimPending,
    pvcDataSourceClone(srcName,
        pvcNamed(dstName, baseClaim())
```